parser/src/ast.rs

```diff
@@ -0,0 +1,103 @@
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct Located<T> {
+    pub l: usize,
+    pub r: usize,
+    pub node: T,
+}
+
+#[derive(Clone, Debug)]
+pub struct Jml<'source> {
+    pub header: Vec<Statement<'source>>,
+    pub body: Expression<'source>,
+}
+
+pub type Statement<'source> = Located<StatementKind<'source>>;
+
+#[derive(Clone, Debug)]
+pub enum StatementKind<'source> {
+    Bind {
+        identifier: Identifier<'source>,
+        expression: Expression<'source>,
+    },
+}
+pub type Identifier<'source> = Located<&'source str>;
+
+pub type Expression<'source> = Located<ExpressionKind<'source>>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Key<'source> {
+    Ident(Identifier<'source>),
+    Expression(Expression<'source>),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ExpressionKind<'source> {
+    Null,
+    Float(f64),
+    Bool(bool),
+    Int(i64),
+    String(&'source str),
```

Comment:
Correct me if I m wrong, but at a later stage when converting this to JmlValue all these str slices are converted to String anyways? Did you consider an option to just store Rc<String> here to avoid the lifetime notation? As far as I understood you have to explicitly specify lifetime because Expression requires it..

Reply:
You're right, when converting to JmlValue, all these &str slices will eventually be converted to String, but it's done only once. However, I avoided using Rc because it enforces borrowing rules at runtime, which could slow down the evaluation. I also prefer using &str slice to copy or clone expressions efficiently since String would need to copy the data on the heap during a clone operation, which is more costly.

Reply:
Speaking of performance - I do not think that there will be any noticeable performance benefit of using &str vs Rc<String> as the slowest parts will be elsewhere. But as you said in other chat there is one case when slice is not converted to owned in the case of Identifier.